### PR TITLE
FAI-7338 Adds sheets source

### PR DESCRIPTION
**Pull request by ypc-faros**
## Description

Pulls from 1) local spreadsheet 2) Google spreadsheet or 3) spreadsheet from an Amazon S3 bucket.
Emits one generic record per row keyed by each header value.

For the table below:

```markdown
| Name    | Age | Country   | Occupation | Salary  |
| ------- | --- | --------- | ---------- | ------- |
| John    | 28  | USA       | Engineer   | 75000   |
| Emily   | 32  | Canada    | Designer   | 62000   |
| Michael | 45  | UK        | Manager    | 95000   |
| Sara    | 23  | Germany   | Student    | 0       |
| Daniel  | 35  | France    | Writer     | 42000   |
| Jessica | 29  | Australia | Doctor     | 120000  |
| Alex    | 31  | Spain     | Teacher    | 52000   |
| Olivia  | 27  | Italy     | Artist     | 38000   |
| William | 40  | Japan     | Scientist  | 80000   |
| Sophia  | 22  | Brazil    | Programmer | 60000   |
```
the source would emit:

```
{"record":{"stream":"sheets","emitted_at":1692370506212,"data":{"Name":"John","Age":28,"Country":"USA","Occupation":"Engineer","Salary":75000}},"type":"RECORD"}
{"record":{"stream":"sheets","emitted_at":1692370506212,"data":{"Name":"Emily","Age":32,"Country":"Canada","Occupation":"Designer","Salary":62000}},"type":"RECORD"}
{"record":{"stream":"sheets","emitted_at":1692370506213,"data":{"Name":"Michael","Age":45,"Country":"UK","Occupation":"Manager","Salary":95000}},"type":"RECORD"}
{"record":{"stream":"sheets","emitted_at":1692370506213,"data":{"Name":"Sara","Age":23,"Country":"Germany","Occupation":"Student","Salary":0}},"type":"RECORD"}
{"record":{"stream":"sheets","emitted_at":1692370506213,"data":{"Name":"Daniel","Age":35,"Country":"France","Occupation":"Writer","Salary":42000}},"type":"RECORD"}
{"record":{"stream":"sheets","emitted_at":1692370506213,"data":{"Name":"Jessica","Age":29,"Country":"Australia","Occupation":"Doctor","Salary":120000}},"type":"RECORD"}
{"record":{"stream":"sheets","emitted_at":1692370506213,"data":{"Name":"Alex","Age":31,"Country":"Spain","Occupation":"Teacher","Salary":52000}},"type":"RECORD"}
{"record":{"stream":"sheets","emitted_at":1692370506213,"data":{"Name":"Olivia","Age":27,"Country":"Italy","Occupation":"Artist","Salary":38000}},"type":"RECORD"}
{"record":{"stream":"sheets","emitted_at":1692370506213,"data":{"Name":"William","Age":40,"Country":"Japan","Occupation":"Scientist","Salary":80000}},"type":"RECORD"}
{"record":{"stream":"sheets","emitted_at":1692370506213,"data":{"Name":"Sophia","Age":22,"Country":"Brazil","Occupation":"Programmer","Salary":60000}},"type":"RECORD"}
```

## Type of change
- [ ] Bug fix
- [X] New feature
- [ ] Breaking change

## Related issues

> Fix [#1]() 

## Migration notes

> Describe migration notes if any

## Extra info

> Add any additional information

### CODE REVIEW

**Review by chalenge — COMMENTED** [review]
lgtm 🚢

File: sources/sheets-source/src/sheets-reader.ts [hunk 1]
```diff
@@ -0,0 +1,307 @@
...
+    logger?: AirbyteLogger
+  ): AsyncGenerator<Dictionary<any, string>> {
+    for (const sheetName of this.getSheetNames()) {
+      logger?.info(`Reading sheet: ${sheetName}`);
+
+      const rows = XLSX.utils.sheet_to_json(this.wb.Sheets[sheetName], {
+        defval: null,
+      });
+
+      for (const row of rows) {
```
Comment (tovbinm) [inline sources/sheets-source/src/sheets-reader.ts:152]: it's a good idea to include the sheet name into the result. how about:
```
yield { sheetName, row }; 
```

File: sources/sheets-source/resources/spec.json [hunk 2]
```diff
@@ -0,0 +1,170 @@
...
+    "title": "Sheets Spec",
+    "type": "object",
+    "required": [
+      "spreadsheet_source"
+    ],
+    "additionalProperties": true,
+    "properties": {
+      "spreadsheet_source": {
+        "order": 1,
+        "title": "Spreadsheet Source",
```
Comment (tovbinm) [inline sources/sheets-source/resources/spec.json:14]: we currently only support `xlsx`. Let's note that somewhere.

File: sources/sheets-source/resources/spec.json [hunk 3]
```diff
@@ -0,0 +1,170 @@
...
+        "title": "Stream Name",
+        "description": "Name of the data stream",
+        "type": "string",
+        "default": "sheets",
+        "examples": [
+          "sales",
+          "customers",
+          "orders"
+        ]
+      }
```
Comment (tovbinm) [inline sources/sheets-source/resources/spec.json:167]: ident + clean spaces

**Comment by cjwooo** [comment]
Please add an [acceptance-test-config](https://github.com/faros-ai/airbyte-connectors/blob/main/sources/circleci-source/acceptance-test-config.yml) along with adding test credentials to the repo's secrets, OR [exclude this source from the acceptance tests](https://github.com/faros-ai/airbyte-connectors/blob/main/.github/workflows/skip-test-sources.txt).

**Review by tovbinm — APPROVED** [review]

File: sources/sheets-source/resources/schemas/sheets.json [hunk 4]
```diff
@@ -2,7 +2,10 @@
   "$schema": "http://json-schema.org/draft-07/schema#",
   "type": "object",
   "properties": {
-    "message": {
+    "sheetName": {
+      "type": "string"
+    },
+    "row": {
       "type": "string"
```
Comment (tovbinm) [inline sources/sheets-source/resources/schemas/sheets.json:9]: `row type should be object, not a string. no?!

Comment (ypc-faros) [reply sources/sheets-source/resources/schemas/sheets.json]: Thanks!!!

**Pull request merged**